sympy/stats/stochastic_process_types.py

```diff
@@ -421,6 +422,20 @@ def _preprocess(self, given_condition, evaluate):
 
         return is_insufficient, trans_probs, state_index, given_condition
 
+    def replace_with_index(self, condition):
+        if isinstance(condition, Relational):
+            if isinstance(condition.rhs, Symbol) or isinstance(condition.rhs, Str):
+                try:
+                    condition = type(condition)(condition.lhs, self.index_of[condition.rhs])
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.rhs)
+            if isinstance(condition.lhs, Symbol) or isinstance(condition.rhs, Str):
+                try:
+                    condition = type(condition)(self.index_of[condition.lhs], condition.rhs)
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.lhs)
```

Comment:
```suggestion
                condition = type(condition)(self.index_of[condition.lhs], condition.rhs)
```

Reply:
I think in the above and this check you are assuming that only one of LHS or RHS will be `RandomIndexedSymbol`?

Reply:
No, actually I'm using 2 ifs, so both of them will be executed if required. Or maybe even better would be - 
``` python
indexes=self.index_of
if isinstance(condition.lhs, Symbol) or isinstance(condition.lhs, Str):
     condition=type(condition)(indexes[condition.lhs],condition.rhs)
if isinstance(condition.rhs, Symbol) or isinstance(condition.rhs, Str):
     condition=type(condition)(condition.lhs,indexes[condition.rhs])
```
The problem with `condition=type(condition)(self.index_of[condition.lhs], self.index_of[condition.rhs])` is that `RandomIndexedSymbol` will not be part of the `indexes` returned by `index_of`

Reply:
For dictionaries there is something called, `get(key, default_value)` which returns `default_value` if `key` isn't there in the dictionary. Can you try using it here?

Reply:
`get` almost works, but fails in one place. If I were to give this - 
Without get - 
``` python
>>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
>>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
>>> P(Eq(Y[3], 0), Eq(Y[1], 1))
0.260000000000000
```
With get - 
``` python
>>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
>>> P(Eq(Y[3], 0), Eq(Y[1], 1))
0.360000000000000
```

Get replaces even integers with their index. I can add an `is_integer` check to solve this. Let me know what you want to do regarding this

Reply:
> ```python
> >>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
> >>> P(Eq(Y[3], 0), Eq(Y[1], 1))
> 0.360000000000000
> ```

This result is correct. The ordering matters after `index_of` was added. So, 1->0, 2->1, 0->2 so, this query is actually, `P(Eq(Y[3], 2), Eq(Y[1], 0))`. The original implementation returns `0.36` as I tried on https://live.sympy.org/

Reply:
Actually, one test fails -
``` python
>>> TS = MatrixSymbol('T', 3, 3)
>>> YS = DiscreteMarkovChain("Y", ['One', 'Two', 3], TS)
>>> assert P(Eq(YS[3], 3), Eq(YS[1], 1)) is S.Zero
```
This fails if I use `get` as get replaces 3 with its index 2, giving probability `T[0, 2]*T[1, 0] + T[1, 1]*T[1, 2] + T[1, 2]*T[2, 2]`.
What should I do?

Reply:
If noticed carefully, `ContinuousMarkovChain` isn't having `index_of` which means we cannot do the same stuff with `ContinuousMarkovChain`. Could you please allow the same features for it?

Reply:
A simple way could be shift `index_of` function in `MarkovProcess` or something like `FiniteStateMarkovProcess`.

Reply:
I see, there are several problems with the way index_of works. Can you please allow `Allow edit from maintainers`? I will be pushing some commits to your branch.

Reply:
Sure, I'll do that

Reply:
Please let me know if I can add anything else